### ISSUE TITLE
shape of state in the output of dynamic_rnn is undetermined

### ISSUE THREAD

**Issue by zakizhou** [tensorflow/tensorflow#8571]
short code:
```
sequence_lengths, (sequences, labels) = bucket_by_sequence_length(
                                              input_length=sequence_length,
                                              tensors=[sequence, label],
                                              batch_size=72,
                                              bucket_boundaries=[80, 160, 240, 320, 400],
                                              dynamic_pad=True)
lstm_cell = core_rnn_cell.BasicLSTMCell(num_units=56)
lookup = tf.nn.embedding_lookup(......)
_, state = tf.nn.dynamic_rnn(
                      cell=lstm_cell,
                      inputs=lookup,
                      sequence_length=sequence_lengths,
                      dtype=tf.float32)
```
then `state.h` and `state.c` should be of shape `[72, 56]`, but actually the first dim(batch_size) is undetermined in program, is it a bug?

**Comment by drpngx** [comment]
CC @ebrevdo for comments

**Comment by ebrevdo** [comment]
What is the shape of lookup?

**Comment by zakizhou** [comment]
@ebrevdo does it matter? let's say embedding matrix is of shape [vocab_size=90000, embed_size=500], so lookup will be of shape [72, unkown, 500], output should be [72, unknown, 56], state.h should be [72, 56]

**Comment by ebrevdo** [comment]
can you confirm lookup is shape [72, unknown, 500]?

**Comment by zakizhou** [comment]
```
import tensorflow as tf
from tensorflow.contrib.rnn import core_rnn_cell
class Option(object):
    def __init__(self, filename, batch_size):
        self.sequences, self.labels, self.sequence_lengths = produce_input(filename=filename, batch_size=batch_size)
        self.vocab_size = 90000
        self.embedding_size = 500
        self.hidden_units = 56
        self.num_classes = 2



option = Option('train', 72)
embedding = tf.get_variable(name='embedding',
                            shape=[option.vocab_size, option.embedding_size],
                            initializer=tf.truncated_normal_initializer(stddev=0.05),
                            dtype=tf.float32)
lookup = tf.nn.embedding_lookup(embedding, option.sequences)
lstm_cell = core_rnn_cell.BasicLSTMCell(num_units=option.hidden_units, state_is_tuple=True)
output, state = tf.nn.dynamic_rnn(cell=lstm_cell,
                                  inputs=lookup,
                                  sequence_length=option.sequence_lengths,
                                  dtype=tf.float32)
output
Out[6]: 
<tf.Tensor 'rnn/transpose:0' shape=(72, ?, 56) dtype=float32>
lookup
Out[7]: 
<tf.Tensor 'embedding_lookup:0' shape=(72, ?, 500) dtype=float32>
state
Out[8]: 
LSTMStateTuple(c=<tf.Tensor 'rnn/while/Exit_2:0' shape=(?, 56) dtype=float32>, h=<tf.Tensor 'rnn/while/Exit_3:0' shape=(?, 56) dtype=float32>)

```

**Comment by girving** [comment]
@ebrevdo What's the status of this issue?

**Comment by ebrevdo** [comment]
I think this is fixed in tf 1.2... OP can you confirm?

**Comment by Alex-Fabbri** [comment]
Hi, I'm having the same problem with the shape of the state being undetermined when using a dynamic_rnn. 
I'm using tensorflow version 1.2.1. 
`        (outputs,final_state) =(
        tf.nn.bidirectional_dynamic_rnn(cell_fw=encoder_cell,
                                    cell_bw=encoder_cell,
                                    inputs=encoder_inputs_embedded,
                                    sequence_length=encoder_inputs_length,
                                    dtype=tf.float32, time_major=True)
        )
`
Encoder_inputs_embedded is the following:
Tensor("embedding_lookup:0", shape=(8, 11, 20), dtype=float32) while final_state is the following:
(LSTMStateTuple(c=<tf.Tensor 'bidirectional_rnn/fw/fw/while/Exit_2:0' shape=(?, 20) dtype=float32>, h=<tf.Tensor 'bidirectional_rnn/fw/fw/while/Exit_3:0' shape=(?, 20) dtype=float32>), LSTMStateTuple(c=<tf.Tensor 'bidirectional_rnn/bw/bw/while/Exit_2:0' shape=(?, 20) dtype=float32>, h=<tf.Tensor 'bidirectional_rnn/bw/bw/while/Exit_3:0' shape=(?, 20) dtype=float32>))

**Comment by ebrevdo** [comment]
What about in tf 1.3?

**Comment by Alex-Fabbri** [comment]
I just tried it on tf 1.3, and it worked. Thanks :)

**Comment by ebrevdo** [comment]
Great!
